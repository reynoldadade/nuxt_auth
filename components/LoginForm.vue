<template>
	<form
		class="flex flex-col pt-3 md:pt-8"
		:class="{ shaker: error }"
		@submit.prevent="login()"
	>
		<div class="flex flex-col pt-4">
			<label for="email" class="text-lg">Email</label>
			<input
				type="email"
				id="email"
				v-model="$v.email.$model"
				placeholder="your@email.com"
				required
				class="shadow appearance-none border rounded w-full py-2 px-3 text-gray-700 mt-1 leading-tight focus:outline-none focus:shadow-outline"
			/>
		</div>

		<div class="flex flex-col pt-4">
			<label for="password" class="text-lg">Password</label>
			<div class="flex">
				<input
					:type="showPassword ? 'text' : 'password'"
					id="password"
					v-model="$v.password.$model"
					placeholder="Password"
					required
					class="shadow appearance-none border rounded w-full py-2 px-3 text-gray-700 mt-1 leading-tight focus:outline-none focus:shadow-outline"
				/>
				<button
					class="py-2 px-3 focus:outline-none"
					@click="showHidePassword"
					type="button"
				>
					<i
						:class="
							showPassword ? 'fas fa-eye' : 'fas fa-eye-slash'
						"
					></i>
				</button>
			</div>
		</div>
		<p class="text-xs text-red-500 font-bold mt-2 text-center" v-if="error">
			<i class="fas fa-exclamation-triangle"></i>
			{{ errorMessage }}
		</p>

		<button
			id="loginButton"
			type="submit"
			:disabled="formValid"
			class="bg-black text-white font-bold text-lg hover:bg-gray-700 p-2 mt-8 disabled:opacity-75 disabled:cursor-not-allowed"
		>
			Log In
			<i class="fas fa-circle-notch fa-spin" v-if="loading"></i>
		</button>
	</form>
</template>

<script>
import { required, email } from 'vuelidate/lib/validators';
import { mapActions } from 'vuex';
export default {
	data() {
		return {
			email: '',
			password: '',
			loading: false,
			error: false,
			errorMessage: 'Incorrect username/password.',
			showPassword: false,
		};
	},
	components: {},
	validations: {
		email: {
			required,
			email,
		},
		password: {
			required,
		},
	},
	computed: {
		formValid() {
			return this.$v.email.$invalid && this.$v.password.$invalid;
		},
	},
	methods: {
		...mapActions({
			storePermissions: 'user/storePermissions',
		}),
		showHidePassword() {
			this.showPassword = !this.showPassword;
		},

		async login() {
			const { redirect_url, expect_token } = this.$route.query;
			this.error = false;
			this.loading = true;
			this.$axios
				.$post('auth/login', {
					email: this.email,
					password: this.password,
				})
				.then(({ data }) => {
					this.loading = false;
					// console.log(data);
					this.$cookies.set('s_token', data.access_token, {
						path: '/',

						maxAge: 60 * 60 * 24 * 7,
						// secure: true,
					});
					mixpanel.track('User Logged', {
						User: data.user.name,
						'Login time': new Date().toLocaleString(),
					});

					mixpanel.identify(data.user.name);

					if (redirect_url) {
						const outURL = `http://${redirect_url}?token=${data.access_token}`;
						return window.location.replace(outURL);
					}
					// this.$cookies.removeAll();
					window.location.replace(
						process.env.WALULEL_LINK +
							'/products?token=' +
							data.access_token
					);
				})
				.catch(({ message, response }) => {
					this.loading = false;
					if (response && response.data) {
						this.errorMessage = response.data.error;
					}
					this.error = true;
				});
		},
	},
};
</script>

<style></style>
