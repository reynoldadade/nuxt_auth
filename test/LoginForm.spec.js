import { mount, createLocalVue, RouterLinkStub } from '@vue/test-utils';
import Vuelidate from 'vuelidate';
import LoginForm from '@/components/LoginForm.vue';

const localVue = createLocalVue();
localVue.use(Vuelidate);
jest.mock('axios', () => ({
	get: Promise.resolve('value'),
}));
describe('LoginForm', () => {
	test('is a Vue instance', () => {
		const wrapper = mount(LoginForm, {
			localVue,
			stubs: { NuxtLink: RouterLinkStub },
		});
		expect(wrapper.exists()).toBeTruthy();
	});
	// it('it fetches async data when button is clicked', () => {
	// 	const wrapper = mount(LoginForm, {
	// 		localVue,
	// 		stubs: { NuxtLink: RouterLinkStub },
	// 	});
	// 	wrapper.find('button#loginButton').trigger('click');
	// 	expect(wrapper.text()).toBe('value');
	// });
});
