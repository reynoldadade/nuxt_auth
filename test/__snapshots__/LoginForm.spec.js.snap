// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm.vue html should render correctly 1`] = `
"<form class=\\"flex flex-col pt-3 md:pt-8\\">
  <div class=\\"flex flex-col pt-4\\"><label for=\\"email\\" class=\\"text-lg\\">Email</label> <input type=\\"email\\" id=\\"email\\" placeholder=\\"your@email.com\\" required=\\"required\\" class=\\"shadow appearance-none border rounded w-full py-2 px-3 text-gray-700 mt-1 leading-tight focus:outline-none focus:shadow-outline\\"></div>
  <div class=\\"flex flex-col pt-4\\"><label for=\\"password\\" class=\\"text-lg\\">Password</label> <input type=\\"password\\" id=\\"password\\" placeholder=\\"Password\\" required=\\"required\\" class=\\"shadow appearance-none border rounded w-full py-2 px-3 text-gray-700 mt-1 leading-tight focus:outline-none focus:shadow-outline\\"></div>
  <!----> <button id=\\"loginButton\\" type=\\"submit\\" disabled=\\"disabled\\" class=\\"bg-black text-white font-bold text-lg hover:bg-gray-700 p-2 mt-8 disabled:opacity-75 disabled:cursor-not-allowed\\">
    Log In
    <i class=\\"fas fa-circle-notch fa-spin\\" style=\\"display: none;\\"></i></button>
</form>"
`;
